File: client/Assets/Scripts/Gameplay/EmojiClashRules.cs
using System;
using System.Collections.Generic;
using System.Linq;
using EmojiWar.Client.Content;

namespace EmojiWar.Client.Gameplay.Clash
{
    public static class EmojiClashRules
    {
        public const int TotalTurns = 5;
        private static readonly int[] TurnValues = { 1, 1, 2, 2, 3 };

        private static readonly Dictionary<string, EmojiClashProfile> Profiles = new(StringComparer.OrdinalIgnoreCase)
        {
            ["fire"] = CreateProfile("fire", "Tempo", 5, 2, 0, 0, 0, new[] { "aggressive", "flame", "projectile", "tempo" }, new[] { "nature", "frozen", "passive", "cleanse" }),
            ["water"] = CreateProfile("water", "Control", 5, 0, 1, 0, 0, new[] { "control", "liquid", "cleanse" }, new[] { "flame", "explosive", "tech" }),
            ["lightning"] = CreateProfile("lightning", "Tempo", 5, 2, 0, 0, 0, new[] { "aggressive", "electric", "projectile", "tempo" }, new[] { "liquid", "flexible" }),
            ["ice"] = CreateProfile("ice", "Control", 5, 0, 0, 0, 0, new[] { "control", "frozen", "defense" }, new[] { "aggressive", "explosive", "elusive" }),
            ["magnet"] = CreateProfile("magnet", "Disrupt", 5, 0, 0, 0, 0, new[] { "disrupt", "tech", "control" }, new[] { "projectile", "electric", "guard" }),
            ["bomb"] = CreateProfile("bomb", "Burst", 6, 1, 0, 0, 0, new[] { "aggressive", "explosive", "projectile" }, new[] { "scaling", "guard", "reactive" }),
            ["mirror"] = CreateProfile("mirror", "Trick", 5, 0, 0, 1, 0, new[] { "reactive", "trick", "flexible" }, new[] { "tempo", "elusive" }),
            ["hole"] = CreateProfile("hole", "Trap", 5, 0, 1, 0, 0, new[] { "disrupt", "void", "trap" }, new[] { "projectile", "tech", "electric" }),
            ["shield"] = CreateProfile("shield", "Defense", 5, 0, 0, 0, 1, new[] { "defense", "guard", "safe" }, new[] { "aggressive", "explosive", "poison" }),
            ["snake"] = CreateProfile("snake", "Pressure", 5, 1, 1, 0, 0, new[] { "poison", "trick", "scaling" }, new[] { "support", "passive", "nature", "reactive" }),
            ["soap"] = CreateProfile("soap", "Cleanse", 5, 0, 0, 0, 1, new[] { "cleanse", "liquid", "control" }, new[] { "poison", "flame", "tech" }),
            ["plant"] = CreateProfile("plant", "Scaling", 5, 0, 1, 1, 0, new[] { "scaling", "nature", "passive" }, new[] { "liquid", "guard", "void" }),
            ["wind"] = CreateProfile("wind", "Flex", 5, 1, 0, 0, 0, new[] { "flexible", "tempo", "control" }, new[] { "explosive", "disrupt", "void" }),
            ["heart"] = CreateProfile("heart", "Comeback", 4, 0, 1, 1, 3, new[] { "support", "comeback", "flexible" }, new[] { "aggressive", "tempo", "late", "elusive" }),
            ["ghost"] = CreateProfile("ghost", "Late", 5, 0, 1, 1, 0, new[] { "late", "elusive", "trick" }, new[] { "guard", "passive" }),
            ["chain"] = CreateProfile("chain", "Bind", 5, 0, 0, 0, 1, new[] { "defense", "bind", "disrupt" }, new[] { "explosive", "poison" })
        };

        private static readonly Dictionary<string, Dictionary<string, int>> HeadToHeadBonuses = new(StringComparer.OrdinalIgnoreCase)
        {
            ["water"] = BuildBonusMap(("fire", 2), ("bomb", 2), ("magnet", 1)),
            ["fire"] = BuildBonusMap(("plant", 2), ("ice", 1), ("snake", 1)),
            ["lightning"] = BuildBonusMap(("water", 2), ("wind", 1), ("soap", 1)),
            ["ice"] = BuildBonusMap(("bomb", 2), ("lightning", 1), ("ghost", 1)),
            ["magnet"] = BuildBonusMap(("lightning", 2), ("fire", 1), ("bomb", 1), ("shield", 1), ("mirror", 1), ("chain", 1)),
            ["bomb"] = BuildBonusMap(("plant", 2), ("shield", 1), ("mirror", 1), ("chain", 1)),
            ["mirror"] = BuildBonusMap(("lightning", 1), ("ghost", 2), ("bomb", 1), ("snake", 1), ("chain", 1)),
            ["hole"] = BuildBonusMap(("lightning", 2), ("magnet", 2), ("bomb", 1), ("mirror", 1)),
            ["shield"] = BuildBonusMap(("fire", 2), ("bomb", 1), ("snake", 2), ("lightning", 1)),
            ["snake"] = BuildBonusMap(("plant", 2), ("heart", 2), ("water", 1), ("mirror", 1)),
            ["soap"] = BuildBonusMap(("snake", 2), ("fire", 1), ("magnet", 2), ("ghost", 1)),
            ["plant"] = BuildBonusMap(("water", 2), ("shield", 1), ("hole", 2)),
            ["wind"] = BuildBonusMap(("bomb", 2), ("magnet", 2), ("hole", 2), ("ghost", 2)),
            ["heart"] = BuildBonusMap(("ghost", 3), ("lightning", 1), ("bomb", 1), ("fire", 1), ("ice", 1)),
            ["ghost"] = BuildBonusMap(("shield", 2), ("plant", 1), ("chain", 1)),
            ["chain"] = BuildBonusMap(("bomb", 2), ("fire", 1), ("wind", 1), ("snake", 1))
        };

        private static readonly Dictionary<string, string> TimingReasonText = new(StringComparer.OrdinalIgnoreCase)
        {
            ["early"] = "hits hard early.",
            ["late"] = "spikes late in the match.",
            ["final"] = "peaks on the final turn.",
            ["behind"] = "gets stronger while behind."
        };

        public static IReadOnlyList<string> LaunchRoster => EmojiIdUtility.LaunchRoster.Select(EmojiIdUtility.ToApiId).ToArray();

        public static int GetTurnValue(int turnIndex)
        {
            if (turnIndex < 0 || turnIndex >= TurnValues.Length)
            {
                throw new ArgumentOutOfRangeException(nameof(turnIndex), turnIndex, "Emoji Clash has exactly five turns.");
            }

            return TurnValues[turnIndex];
        }

        public static EmojiClashProfile GetUnitProfile(string unitKey)
        {
            var normalizedKey = NormalizeUnitKey(unitKey);
            if (!Profiles.TryGetValue(normalizedKey, out var profile))
            {
                throw new ArgumentOutOfRangeException(nameof(unitKey), unitKey, "Unknown Emoji Clash unit.");
            }

            return profile;
        }

        public static EmojiClashResolvedTurn ResolveTurn(EmojiClashMatchState state, string playerUnitKey, string opponentUnitKey)
        {
            if (state == null)
            {
                throw new ArgumentNullException(nameof(state));
            }

            var turnIndex = Math.Clamp(state.CurrentTurnIndex, 0, TotalTurns - 1);
            var turnNumber = turnIndex + 1;
            var playerProfile = GetUnitProfile(playerUnitKey);
            var opponentProfile = GetUnitProfile(opponentUnitKey);
            var playerBreakdown = BuildBreakdown(playerProfile, opponentProfile, turnNumber, state.PlayerScore, state.OpponentScore);
            var opponentBreakdown = BuildBreakdown(opponentProfile, playerProfile, turnNumber, state.OpponentScore, state.PlayerScore);
            var outcome = ResolveOutcome(playerBreakdown.TotalPower, opponentBreakdown.TotalPower);

            return new EmojiClashResolvedTurn
            {
                TurnNumber = turnNumber,
                TurnValue = GetTurnValue(turnIndex),
                Outcome = outcome,
                PlayerUnitKey = playerProfile.UnitKey,
                OpponentUnitKey = opponentProfile.UnitKey,
                PlayerBreakdown = playerBreakdown,
                OpponentBreakdown = opponentBreakdown,
                PlayerFacingReason = GetReadableReason(playerProfile, opponentProfile, playerBreakdown, opponentBreakdown, outcome)
            };
        }

        public static string GetReadableReason(EmojiClashResolvedTurn resolvedTurn)
        {
            if (resolvedTurn == null)
            {
                return string.Empty;
            }

            return resolvedTurn.PlayerFacingReason ?? string.Empty;
        }

        public static string BuildTurnSummary(EmojiClashTurnRecord record)
        {
            if (record == null)
            {
                return string.Empty;
            }

            var playerName = ToDisplayName(record.PlayerUnitKey);
            var opponentName = ToDisplayName(record.OpponentUnitKey);
            return record.Outcome switch
            {
                EmojiClashTurnOutcome.PlayerWin => $"T{record.TurnNumber} +{record.TurnValue}  {playerName} beat {opponentName}",
                EmojiClashTurnOutcome.OpponentWin => $"T{record.TurnNumber} +{record.TurnValue}  {playerName} lost to {opponentName}",
                _ => $"T{record.TurnNumber} +{record.TurnValue}  Draw"
            };
        }

        public static string ToDisplayName(string unitKey)
        {
            var normalizedKey = NormalizeUnitKey(unitKey);
            if (EmojiIdUtility.TryFromApiId(normalizedKey, out var emojiId))
            {
                return EmojiIdUtility.ToDisplayName(emojiId);
            }

            return string.IsNullOrWhiteSpace(unitKey)
                ? "Unit"
                : char.ToUpperInvariant(normalizedKey[0]) + normalizedKey[1..];
        }

        public static string NormalizeUnitKey(string unitKey)
        {
            if (string.IsNullOrWhiteSpace(unitKey))
            {
                return "unknown";
            }

            var key = unitKey.Trim().ToLowerInvariant();
            var dot = key.LastIndexOf('.');
            if (dot >= 0 && dot < key.Length - 1)
            {
                key = key[(dot + 1)..];
            }

            key = key.Replace("emoji_", string.Empty)
                .Replace("emoji-", string.Empty)
                .Replace("unit_", string.Empty)
                .Replace("unit-", string.Empty)
                .Replace(" ", string.Empty)
                .Replace("_", string.Empty)
                .Replace("-", string.Empty);

            return string.IsNullOrWhiteSpace(key) ? "unknown" : key;
        }

        private static EmojiClashTurnOutcome ResolveOutcome(int playerPower, int opponentPower)
        {
            if (playerPower > opponentPower)
            {
                return EmojiClashTurnOutcome.PlayerWin;
            }

            if (opponentPower > playerPower)
            {
                return EmojiClashTurnOutcome.OpponentWin;
            }

            return EmojiClashTurnOutcome.Draw;
        }

        private static EmojiClashPowerBreakdown BuildBreakdown(
            EmojiClashProfile profile,
            EmojiClashProfile opponent,
            int turnNumber,
            int sideScore,
            int otherScore)
        {
            var matchupBonus = ComputeMatchupBonus(profile, opponent);
            var timingBonus = ComputeTimingBonus(profile, turnNumber);
            var behindBonus = sideScore < otherScore ? profile.BehindBonus : 0;
            var totalPower = profile.BasePower + matchupBonus + timingBonus + behindBonus;

            return new EmojiClashPowerBreakdown
            {
                UnitKey = profile.UnitKey,
                BasePower = profile.BasePower,
                MatchupBonus = matchupBonus,
                TimingBonus = timingBonus,
                BehindBonus = behindBonus,
                TotalPower = totalPower,
                PrimaryReason = ResolvePrimaryReason(profile, opponent, matchupBonus, timingBonus, behindBonus, turnNumber)
            };
        }

        private static int ComputeMatchupBonus(EmojiClashProfile profile, EmojiClashProfile opponent)
        {
            var counterMatches = opponent.Tags
                .Where(tag => profile.CounterTags.Contains(tag, StringComparer.OrdinalIgnoreCase))
                .Distinct(StringComparer.OrdinalIgnoreCase)
                .Count();

            var tagBonus = Math.Min(2, counterMatches);
            var directBonus = 0;
            if (HeadToHeadBonuses.TryGetValue(profile.UnitKey, out var directMap) &&
                directMap.TryGetValue(opponent.UnitKey, out var bonus))
            {
                directBonus = bonus;
            }

            return tagBonus + directBonus;
        }

        private static int ComputeTimingBonus(EmojiClashProfile profile, int turnNumber)
        {
            var total = 0;
            if (turnNumber <= 2)
            {
                total += profile.EarlyBonus;
            }

            if (turnNumber >= 4)
            {
                total += profile.LateBonus;
            }

            if (turnNumber == TotalTurns)
            {
                total += profile.FinalTurnBonus;
            }

            return total;
        }

        private static string ResolvePrimaryReason(
            EmojiClashProfile profile,
            EmojiClashProfile opponent,
            int matchupBonus,
            int timingBonus,
            int behindBonus,
            int turnNumber)
        {
            if (matchupBonus >= Math.Max(timingBonus, behindBonus) && matchupBonus > 0)
            {
                return $"{profile.DisplayName} countered {opponent.DisplayName.ToLowerInvariant()}.";
            }

            if (turnNumber == TotalTurns && profile.FinalTurnBonus > 0 && timingBonus > 0)
            {
                return $"{profile.DisplayName} {TimingReasonText["final"]}";
            }

            if (turnNumber >= 4 && profile.LateBonus > 0 && timingBonus > 0)
            {
                return $"{profile.DisplayName} {TimingReasonText["late"]}";
            }

            if (turnNumber <= 2 && profile.EarlyBonus > 0 && timingBonus > 0)
            {
                return $"{profile.DisplayName} {TimingReasonText["early"]}";
            }

            if (behindBonus > 0)
            {
                return $"{profile.DisplayName} {TimingReasonText["behind"]}";
            }

            return $"{profile.DisplayName} won the cleaner clash.";
        }

        private static string GetReadableReason(
            EmojiClashProfile playerProfile,
            EmojiClashProfile opponentProfile,
            EmojiClashPowerBreakdown playerBreakdown,
            EmojiClashPowerBreakdown opponentBreakdown,
            EmojiClashTurnOutcome outcome)
        {
            return outcome switch
            {
                EmojiClashTurnOutcome.PlayerWin => playerBreakdown.PrimaryReason,
                EmojiClashTurnOutcome.OpponentWin => opponentBreakdown.PrimaryReason,
                _ when playerBreakdown.MatchupBonus > 0 && opponentBreakdown.MatchupBonus > 0 => $"{playerProfile.DisplayName} and {opponentProfile.DisplayName} traded clean counters.",
                _ => $"{playerProfile.DisplayName} and {opponentProfile.DisplayName} finished even. No points awarded."
            };
        }

        private static EmojiClashProfile CreateProfile(
            string unitKey,
            string role,
            int basePower,
            int earlyBonus,
            int lateBonus,
            int finalTurnBonus,
            int behindBonus,
            string[] tags,
            string[] counters)
        {
            return new EmojiClashProfile
            {
                UnitKey = unitKey,
                DisplayName = ToDisplayName(unitKey),
                BasePower = basePower,
                Role = role,
                Tags = tags ?? Array.Empty<string>(),
                CounterTags = counters ?? Array.Empty<string>(),
                EarlyBonus = earlyBonus,
                LateBonus = lateBonus,
                FinalTurnBonus = finalTurnBonus,
                BehindBonus = behindBonus
            };
        }

        private static Dictionary<string, int> BuildBonusMap(params (string unitKey, int bonus)[] entries)
        {
            var map = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
            foreach (var entry in entries)
            {
                map[entry.unitKey] = entry.bonus;
            }

            return map;
        }
    }
}

File: supabase/functions/_shared/emoji-clash-rules.ts
import type { EmojiId, Winner } from "./contracts.ts";

export const EMOJI_CLASH_RULES_VERSION = "emoji-clash-pvp-001";
export const EMOJI_CLASH_TOTAL_TURNS = 5;
export const EMOJI_CLASH_TURN_VALUES = [1, 1, 2, 2, 3] as const;
export const EMOJI_CLASH_PICK_TTL_SECONDS = 15;
export const EMOJI_CLASH_QUEUE_TTL_SECONDS = 120;
export const EMOJI_CLASH_TIMEOUT_FORFEIT_STRIKES = 2;

export type EmojiClashTurnOutcome = "player_a" | "player_b" | "draw";

export interface EmojiClashProfile {
  unitKey: EmojiId;
  displayName: string;
  role: string;
  basePower: number;
  tags: string[];
  counterTags: string[];
  earlyBonus: number;
  lateBonus: number;
  finalTurnBonus: number;
  behindBonus: number;
}

export interface EmojiClashTurnRecord {
  turnNumber: number;
  turnValue: number;
  playerAUnitKey: EmojiId | "";
  playerBUnitKey: EmojiId | "";
  playerATimedOut?: boolean;
  playerBTimedOut?: boolean;
  playerATimeoutBurnUnitKey?: EmojiId | "";
  playerBTimeoutBurnUnitKey?: EmojiId | "";
  playerACombatPower: number;
  playerBCombatPower: number;
  outcome: EmojiClashTurnOutcome;
  playerAScoreAfter: number;
  playerBScoreAfter: number;
  reason: string;
}

export interface EmojiClashPublicState {
  phase: "queue" | "pick" | "finished" | "queue_cancelled";
  queueTicket?: string;
  opponentType?: "human" | "bot_fill";
  botProfileId?: string;
  botFillReason?: string;
  opponentDisplayName?: string;
  opponentAvatarKey?: string;
  currentTurnIndex: number;
  playerAScore: number;
  playerBScore: number;
  playerAUsedUnits: EmojiId[];
  playerBUsedUnits: EmojiId[];
  matchmakingRatingAtQueue?: number;
  queueStartedAt?: string;
  queueExpiresAt?: string;
  playerATimeoutStrikes?: number;
  playerBTimeoutStrikes?: number;
  turnHistory: EmojiClashTurnRecord[];
  turnDeadlineAt?: string;
  winner?: Winner | null;
  finishReason?: string;
  forfeitSides?: Array<"player_a" | "player_b">;
  systemNote?: string;
  cancelledAt?: string;
  cancelledBy?: string;
}

export const EMOJI_CLASH_ROSTER: EmojiId[] = [
  "fire",
  "water",
  "lightning",
  "ice",
  "magnet",
  "bomb",
  "mirror",
  "hole",
  "shield",
  "snake",
  "soap",
  "plant",
  "wind",
  "heart",
  "ghost",
  "chain",
];

const profiles: Record<EmojiId, EmojiClashProfile> = {
  fire: createProfile("fire", "Tempo", 5, 2, 0, 0, 0, ["aggressive", "flame", "projectile", "tempo"], ["nature", "frozen", "passive", "cleanse"]),
  water: createProfile("water", "Control", 5, 0, 1, 0, 0, ["control", "liquid", "cleanse"], ["flame", "explosive", "tech"]),
  lightning: createProfile("lightning", "Tempo", 5, 2, 0, 0, 0, ["aggressive", "electric", "projectile", "tempo"], ["liquid", "flexible"]),
  ice: createProfile("ice", "Control", 5, 0, 0, 0, 0, ["control", "frozen", "defense"], ["aggressive", "explosive", "elusive"]),
  magnet: createProfile("magnet", "Disrupt", 5, 0, 0, 0, 0, ["disrupt", "tech", "control"], ["projectile", "electric", "guard"]),
  bomb: createProfile("bomb", "Burst", 6, 1, 0, 0, 0, ["aggressive", "explosive", "projectile"], ["scaling", "guard", "reactive"]),
  mirror: createProfile("mirror", "Trick", 5, 0, 0, 1, 0, ["reactive", "trick", "flexible"], ["tempo", "elusive"]),
  hole: createProfile("hole", "Trap", 5, 0, 1, 0, 0, ["disrupt", "void", "trap"], ["projectile", "tech", "electric"]),
  shield: createProfile("shield", "Defense", 5, 0, 0, 0, 1, ["defense", "guard", "safe"], ["aggressive", "explosive", "poison"]),
  snake: createProfile("snake", "Pressure", 5, 1, 1, 0, 0, ["poison", "trick", "scaling"], ["support", "passive", "nature", "reactive"]),
  soap: createProfile("soap", "Cleanse", 5, 0, 0, 0, 1, ["cleanse", "liquid", "control"], ["poison", "flame", "tech"]),
  plant: createProfile("plant", "Scaling", 5, 0, 1, 1, 0, ["scaling", "nature", "passive"], ["liquid", "guard", "void"]),
  wind: createProfile("wind", "Flex", 5, 1, 0, 0, 0, ["flexible", "tempo", "control"], ["explosive", "disrupt", "void"]),
  heart: createProfile("heart", "Comeback", 4, 0, 1, 1, 3, ["support", "comeback", "flexible"], ["aggressive", "tempo", "late", "elusive"]),
  ghost: createProfile("ghost", "Late", 5, 0, 1, 1, 0, ["late", "elusive", "trick"], ["guard", "passive"]),
  chain: createProfile("chain", "Bind", 5, 0, 0, 0, 1, ["defense", "bind", "disrupt"], ["explosive", "poison"]),
};

const headToHead: Partial<Record<EmojiId, Partial<Record<EmojiId, number>>>> = {
  water: bonus([["fire", 2], ["bomb", 2], ["magnet", 1]]),
  fire: bonus([["plant", 2], ["ice", 1], ["snake", 1]]),
  lightning: bonus([["water", 2], ["wind", 1], ["soap", 1]]),
  ice: bonus([["bomb", 2], ["lightning", 1], ["ghost", 1]]),
  magnet: bonus([["lightning", 2], ["fire", 1], ["bomb", 1], ["shield", 1], ["mirror", 1], ["chain", 1]]),
  bomb: bonus([["plant", 2], ["shield", 1], ["mirror", 1], ["chain", 1]]),
  mirror: bonus([["lightning", 1], ["ghost", 2], ["bomb", 1], ["snake", 1], ["chain", 1]]),
  hole: bonus([["lightning", 2], ["magnet", 2], ["bomb", 1], ["mirror", 1]]),
  shield: bonus([["fire", 2], ["bomb", 1], ["snake", 2], ["lightning", 1]]),
  snake: bonus([["plant", 2], ["heart", 2], ["water", 1], ["mirror", 1]]),
  soap: bonus([["snake", 2], ["fire", 1], ["magnet", 2], ["ghost", 1]]),
  plant: bonus([["water", 2], ["shield", 1], ["hole", 2]]),
  wind: bonus([["bomb", 2], ["magnet", 2], ["hole", 2], ["ghost", 2]]),
  heart: bonus([["ghost", 3], ["lightning", 1], ["bomb", 1], ["fire", 1], ["ice", 1]]),
  ghost: bonus([["shield", 2], ["plant", 1], ["chain", 1]]),
  chain: bonus([["bomb", 2], ["fire", 1], ["wind", 1], ["snake", 1]]),
};

function createProfile(
  unitKey: EmojiId,
  role: string,
  basePower: number,
  earlyBonus: number,
  lateBonus: number,
  finalTurnBonus: number,
  behindBonus: number,
  tags: string[],
  counterTags: string[],
): EmojiClashProfile {
  return {
    unitKey,
    displayName: toDisplayName(unitKey),
    role,
    basePower,
    tags,
    counterTags,
    earlyBonus,
    lateBonus,
    finalTurnBonus,
    behindBonus,
  };
}

function bonus(entries: Array<[EmojiId, number]>): Partial<Record<EmojiId, number>> {
  return Object.fromEntries(entries) as Partial<Record<EmojiId, number>>;
}

export function isEmojiClashUnit(value: string): value is EmojiId {
  return EMOJI_CLASH_ROSTER.includes(value as EmojiId);
}

export function toDisplayName(unitKey: string): string {
  return unitKey.length === 0 ? "No pick" : unitKey.charAt(0).toUpperCase() + unitKey.slice(1);
}

export function getTurnValue(turnIndex: number): number {
  if (turnIndex < 0 || turnIndex >= EMOJI_CLASH_TURN_VALUES.length) {
    throw new Error("Emoji Clash has exactly five turns.");
  }

  return EMOJI_CLASH_TURN_VALUES[turnIndex];
}

export function createInitialClashState(queueTicket: string): EmojiClashPublicState {
  return {
    phase: "queue",
    queueTicket,
    currentTurnIndex: 0,
    playerAScore: 0,
    playerBScore: 0,
    playerAUsedUnits: [],
    playerBUsedUnits: [],
    playerATimeoutStrikes: 0,
    playerBTimeoutStrikes: 0,
    turnHistory: [],
  };
}

export function startPickPhase(state: EmojiClashPublicState, now = new Date()): EmojiClashPublicState {
  return {
    ...state,
    phase: "pick",
    turnDeadlineAt: new Date(now.getTime() + EMOJI_CLASH_PICK_TTL_SECONDS * 1000).toISOString(),
  };
}

export function resolvePickedTurn(
  state: EmojiClashPublicState,
  playerAUnitKey: EmojiId,
  playerBUnitKey: EmojiId,
): EmojiClashPublicState {
  const turnIndex = Math.max(0, Math.min(EMOJI_CLASH_TOTAL_TURNS - 1, state.currentTurnIndex));
  const turnNumber = turnIndex + 1;
  const turnValue = getTurnValue(turnIndex);
  const playerAPower = computePower(playerAUnitKey, playerBUnitKey, turnNumber, state.playerAScore, state.playerBScore);
  const playerBPower = computePower(playerBUnitKey, playerAUnitKey, turnNumber, state.playerBScore, state.playerAScore);
  const outcome: EmojiClashTurnOutcome = playerAPower > playerBPower ? "player_a" : playerBPower > playerAPower ? "player_b" : "draw";
  return appendResolvedTurn(state, {
    turnNumber,
    turnValue,
    playerAUnitKey,
    playerBUnitKey,
    playerACombatPower: playerAPower,
    playerBCombatPower: playerBPower,
    outcome,
    playerAScoreAfter: state.playerAScore + (outcome === "player_a" ? turnValue : 0),
    playerBScoreAfter: state.playerBScore + (outcome === "player_b" ? turnValue : 0),
    reason: buildReason(playerAUnitKey, playerBUnitKey, playerAPower, playerBPower, outcome),
  }, playerAUnitKey, playerBUnitKey);
}

export function resolveTimeoutTurn(
  state: EmojiClashPublicState,
  playerAUnitKey: EmojiId | "",
  playerBUnitKey: EmojiId | "",
  matchId = state.queueTicket ?? "",
  playerAUserId = "player_a",
  playerBUserId = "player_b",
): EmojiClashPublicState {
  const turnIndex = Math.max(0, Math.min(EMOJI_CLASH_TOTAL_TURNS - 1, state.currentTurnIndex));
  const turnNumber = turnIndex + 1;
  const turnValue = getTurnValue(turnIndex);
  const playerATimedOut = !playerAUnitKey;
  const playerBTimedOut = !playerBUnitKey;
  const playerATimeoutBurn = playerATimedOut
    ? selectTimeoutBurnUnit(state.playerAUsedUnits, matchId, turnNumber, "player_a", playerAUserId)
    : "";
  const playerBTimeoutBurn = playerBTimedOut
    ? selectTimeoutBurnUnit(state.playerBUsedUnits, matchId, turnNumber, "player_b", playerBUserId)
    : "";
  const consumedA = playerAUnitKey || playerATimeoutBurn;
  const consumedB = playerBUnitKey || playerBTimeoutBurn;
  const outcome: EmojiClashTurnOutcome = playerAUnitKey && !playerBUnitKey
    ? "player_a"
    : playerBUnitKey && !playerAUnitKey
      ? "player_b"
      : "draw";
  const nextTimeoutStrikesA = (state.playerATimeoutStrikes ?? 0) + (playerATimedOut ? 1 : 0);
  const nextTimeoutStrikesB = (state.playerBTimeoutStrikes ?? 0) + (playerBTimedOut ? 1 : 0);
  const reason = buildTimeoutReason(playerAUnitKey, playerBUnitKey, playerATimeoutBurn, playerBTimeoutBurn, outcome, turnValue);
  return appendResolvedTurn(state, {
    turnNumber,
    turnValue,
    playerAUnitKey: playerAUnitKey || playerATimeoutBurn,
    playerBUnitKey: playerBUnitKey || playerBTimeoutBurn,
    playerATimedOut,
    playerBTimedOut,
    playerATimeoutBurnUnitKey: playerATimeoutBurn,
    playerBTimeoutBurnUnitKey: playerBTimeoutBurn,
    playerACombatPower: playerAUnitKey ? profiles[playerAUnitKey].basePower : 0,
    playerBCombatPower: playerBUnitKey ? profiles[playerBUnitKey].basePower : 0,
    outcome,
    playerAScoreAfter: state.playerAScore + (outcome === "player_a" ? turnValue : 0),
    playerBScoreAfter: state.playerBScore + (outcome === "player_b" ? turnValue : 0),
    reason,
  }, consumedA || null, consumedB || null, {
    playerATimeoutStrikes: nextTimeoutStrikesA,
    playerBTimeoutStrikes: nextTimeoutStrikesB,
  });
}

function appendResolvedTurn(
  state: EmojiClashPublicState,
  record: EmojiClashTurnRecord,
  consumedA: EmojiId | null,
  consumedB: EmojiId | null,
  strikeUpdate?: { playerATimeoutStrikes: number; playerBTimeoutStrikes: number },
): EmojiClashPublicState {
  const nextTurnIndex = record.turnNumber >= EMOJI_CLASH_TOTAL_TURNS ? EMOJI_CLASH_TOTAL_TURNS : record.turnNumber;
  const playerATimeoutStrikes = strikeUpdate?.playerATimeoutStrikes ?? state.playerATimeoutStrikes ?? 0;
  const playerBTimeoutStrikes = strikeUpdate?.playerBTimeoutStrikes ?? state.playerBTimeoutStrikes ?? 0;
  const forfeitSides: Array<"player_a" | "player_b"> = [];
  if (playerATimeoutStrikes >= EMOJI_CLASH_TIMEOUT_FORFEIT_STRIKES) forfeitSides.push("player_a");
  if (playerBTimeoutStrikes >= EMOJI_CLASH_TIMEOUT_FORFEIT_STRIKES) forfeitSides.push("player_b");
  const isForfeit = forfeitSides.length > 0;
  const isFinished = isForfeit || nextTurnIndex >= EMOJI_CLASH_TOTAL_TURNS;
  const nextState: EmojiClashPublicState = {
    ...state,
    currentTurnIndex: nextTurnIndex,
    playerAScore: record.playerAScoreAfter,
    playerBScore: record.playerBScoreAfter,
    playerAUsedUnits: consumedA ? [...state.playerAUsedUnits, consumedA] : [...state.playerAUsedUnits],
    playerBUsedUnits: consumedB ? [...state.playerBUsedUnits, consumedB] : [...state.playerBUsedUnits],
    playerATimeoutStrikes,
    playerBTimeoutStrikes,
    turnHistory: [...state.turnHistory, record],
    phase: isFinished ? "finished" : "pick",
    turnDeadlineAt: isFinished ? "" : new Date(Date.now() + EMOJI_CLASH_PICK_TTL_SECONDS * 1000).toISOString(),
  };
  if (isForfeit) {
    nextState.forfeitSides = forfeitSides;
    nextState.finishReason = forfeitSides.length === 2
      ? "both_timeout_forfeit"
      : `${forfeitSides[0]}_timeout_forfeit`;
    nextState.winner = resolveForfeitWinner(nextState, forfeitSides);
  } else if (isFinished) {
    nextState.finishReason = "completed";
    nextState.winner = nextState.playerAScore > nextState.playerBScore
      ? "player_a"
      : nextState.playerBScore > nextState.playerAScore
        ? "player_b"
        : "draw";
  }

  return nextState;
}

function resolveForfeitWinner(state: EmojiClashPublicState, forfeitSides: Array<"player_a" | "player_b">): Winner {
  if (forfeitSides.length === 1) {
    return forfeitSides[0] === "player_a" ? "player_b" : "player_a";
  }

  if (state.playerAScore > state.playerBScore) return "player_a";
  if (state.playerBScore > state.playerAScore) return "player_b";
  return "draw";
}

export function selectTimeoutBurnUnit(
  usedUnits: EmojiId[],
  matchId: string,
  turnNumber: number,
  playerSide: "player_a" | "player_b",
  userId: string,
): EmojiId | "" {
  const used = new Set(usedUnits);
  const remaining = [...EMOJI_CLASH_ROSTER].filter((unit) => !used.has(unit)).sort();
  if (remaining.length === 0) {
    return "";
  }

  const seed = stableHash(`${matchId}|${turnNumber}|${playerSide}|${userId}`);
  return remaining[seed % remaining.length];
}

function stableHash(value: string): number {
  let hash = 2166136261;
  for (let index = 0; index < value.length; index++) {
    hash ^= value.charCodeAt(index);
    hash = Math.imul(hash, 16777619);
  }

  return hash >>> 0;
}

function buildTimeoutReason(
  playerAUnitKey: EmojiId | "",
  playerBUnitKey: EmojiId | "",
  playerABurn: EmojiId | "",
  playerBBurn: EmojiId | "",
  outcome: EmojiClashTurnOutcome,
  turnValue: number,
): string {
  if (!playerAUnitKey && !playerBUnitKey) {
    return `Both players missed the turn. No points awarded. Player A lost ${toDisplayName(playerABurn)}. Player B lost ${toDisplayName(playerBBurn)}.`;
  }

  if (outcome === "player_a") {
    return `Player B missed the turn. Player A gained +${turnValue}. Player B lost ${toDisplayName(playerBBurn)}.`;
  }

  if (outcome === "player_b") {
    return `Player A missed the turn. Player B gained +${turnValue}. Player A lost ${toDisplayName(playerABurn)}.`;
  }

  return "Turn timer expired.";
}

function computePower(self: EmojiId, opponent: EmojiId, turnNumber: number, ownScore: number, opponentScore: number): number {
  const profile = profiles[self];
  return profile.basePower +
    computeMatchupBonus(profile, profiles[opponent]) +
    computeTimingBonus(profile, turnNumber) +
    (ownScore < opponentScore ? profile.behindBonus : 0);
}

function computeMatchupBonus(profile: EmojiClashProfile, opponent: EmojiClashProfile): number {
  const direct = headToHead[profile.unitKey]?.[opponent.unitKey] ?? 0;
  const tagBonus = opponent.tags.some((tag) => profile.counterTags.includes(tag)) ? 1 : 0;
  return direct + tagBonus;
}

function computeTimingBonus(profile: EmojiClashProfile, turnNumber: number): number {
  let total = 0;
  if (turnNumber <= 2) total += profile.earlyBonus;
  if (turnNumber >= 4) total += profile.lateBonus;
  if (turnNumber === EMOJI_CLASH_TOTAL_TURNS) total += profile.finalTurnBonus;
  return total;
}

function buildReason(playerAUnitKey: EmojiId, playerBUnitKey: EmojiId, powerA: number, powerB: number, outcome: EmojiClashTurnOutcome): string {
  if (outcome === "draw") {
    return `${toDisplayName(playerAUnitKey)} and ${toDisplayName(playerBUnitKey)} fought to a draw.`;
  }

  const winner = outcome === "player_a" ? playerAUnitKey : playerBUnitKey;
  const loser = outcome === "player_a" ? playerBUnitKey : playerAUnitKey;
  const margin = Math.abs(powerA - powerB);
  return margin >= 2
    ? `${toDisplayName(winner)} countered ${toDisplayName(loser).toLowerCase()}.`
    : `${toDisplayName(winner)} won the cleaner clash.`;
}
